src/pywws/service/mqtt.py

```diff
@@ -147,35 +179,65 @@ class ToService(pywws.service.LiveDataService):
-#wind_gust    '"wind_gust"   : "%.2f",' '' 'wind_mph(x)'#
+#idx          '"idx"             : "%Y-%m-%d %H:%M:%S",'#
+#wind_dir     '"wind_dir_degrees": "%.d",' '' 'winddir_degrees(x)'#
+#wind_dir     '"wind_dir_text"   : "%s",' '' 'winddir_text(x)'#
+#wind_ave     '"wind_ave_mps"    : "%.2f",'#
+#wind_ave     '"wind_ave_mph"    : "%.2f",' '' 'wind_mph(x)'#
+#wind_gust    '"wind_gust_mps"   : "%.2f",'#
+#wind_gust    '"wind_gust_mph"   : "%.2f",' '' 'wind_mph(x)'#
+#calc 'wind_chill(data["temp_out"],data["wind_ave"])'         '"wind_chill_c" : "%.1f",'#
+#calc 'temp_f(wind_chill(data["temp_out"],data["wind_ave"]))' '"wind_chill_f" : "%.1f",'#
+#calc 'dew_point(data["temp_out"],data["hum_out"])'           '"dew_point_c" : "%.1f",'#
+#calc 'temp_f(dew_point(data["temp_out"],data["hum_out"]))'   '"dew_point_f" : "%.1f",'#
 #hum_out      '"hum_out"     : "%.d",'#
 #hum_in       '"hum_in"      : "%.d",'#
 #temp_in      '"temp_in_c"   : "%.1f",'#
 #temp_in      '"temp_in_f"   : "%.1f",' '' 'temp_f(x)'#
 #temp_out     '"temp_out_c"  : "%.1f",'#
 #temp_out     '"temp_out_f"  : "%.1f",' '' 'temp_f(x)'#
-#rel_pressure '"rel_pressure": "%.4f",' '' 'pressure_inhg(x)'#
-#calc 'rain_inch(rain_hour(data))' '"rainin": "%g",'#
-#calc 'rain_inch(rain_day(data))' '"dailyrainin": "%g",'#
-#calc 'rain_hour(data)' '"rain": "%g",'#
-#calc 'rain_day(data)' '"dailyrain": "%g",'#
-
+#calc 'apparent_temp(data["temp_out"],data["hum_out"],data["wind_ave"])'         '"temp_out_realfeel_c" : "%.1f",'#
+#calc 'temp_f(apparent_temp(data["temp_out"],data["hum_out"],data["wind_ave"]))' '"temp_out_realfeel_f" : "%.1f",'#
+#rel_pressure '"pressure_rel_hpa" : "%.1f",'#
+#rel_pressure '"pressure_rel_inhg": "%.4f",' '' 'pressure_inhg(x)'#
+#abs_pressure '"pressure_abs_hpa" : "%.1f",'#
+#abs_pressure '"pressure_abs_inhg": "%.4f",' '' 'pressure_inhg(x)'#
+#rain         '"rain_mm"     : "%.1f",'#
+#rain         '"rain_in"     : "%.2f",' '' 'rain_inch(x)'#
+#calc 'rain_hour(data)'            '"rain_last_hour_mm": "%.1f",'#
+#calc 'rain_inch(rain_hour(data))' '"rain_last_hour_in": "%.2f",'#
+#calc 'rain_24hr(data)'            '"rain_last_24hours_mm": "%.1f",'#
+#calc 'rain_inch(rain_24hr(data))' '"rain_last_24hours_in": "%.2f",'#
+#calc 'rain_day(data)'             '"rain_day_mm": "%.1f",'#
+#calc 'rain_inch(rain_day(data))'  '"rain_day_in": "%.2f",'#
+"""
+    template_3080_add = """
+#illuminance  '"illuminance_lux" : "%.1f",'#
+#illuminance  '"illuminance_wm2" : "%.2f",' '' 'illuminance_wm2(x)'#
+#uv           '"uv"          : "%.d",'#
 """
 
     def __init__(self, context, check_params=True):
         super(ToService, self).__init__(context, check_params)
+        # extend template
+        if context.params.get('config', 'ws type') == '3080':
+            self.template += self.template_3080_add
         # get template text
         template = literal_eval(context.params.get(
-            service_name, 'template_txt', pprint.pformat(self.template)))
```

Comment:
What was wrong with pprint.pformat?

Reply:
the whole template ended up on one line, if defaulted, i.e.:

template_txt = '\n#idx          \'"idx"             : "%Y-%m-%d %H:%M:%S",\'#\n#wind_dir     \'"wind_dir_degrees": "%.d",\' \'\' \'winddir_degrees(x)\'#\n#wind_dir     \'"wind_dir_text"   : "%s",\' \'\' \'winddir_text(x)\'#'

which makes it difficult to change. The new version will format it as (in the .ini file):

template_txt = ('#idx          \'"idx"             : "%Y-%m-%d %H:%M:%S",\'#\n'
    '#wind_dir     \'"wind_dir_degrees": "%.d",\' \'\' \'winddir_degrees(x)\'#\n'
    '#wind_dir     \'"wind_dir_text"   : "%s",\' \'\' \'winddir_text(x)\'#'
    )

that is format the documentation uses. This is one value per line and also follows the format used in the source code to specify default.

I removed the template_txt from the example in documentation, to encourage people to use the default and make it easier to start (example is smaller), but then felt that if the resulting format of defaulted template_txt was difficult to change, users may not be happy about it. This way it actually makes it easier, default example is small, template_txt gets defaulted for you, and if any changes are needed later (new values or just cleanup) one can do it easier this way.

pprint.format is still used, but for each line separately.

Reply:
OK, I thought pprint had given me the right thing when I tried it, but I might have misremembered. I wonder if it can be configured to give one item per line, which I agree is what we want.

Reply:
I checked but did not see anything there. From python perspective it is just one string and we want the content formatted special way, taking each line, escaping it and preserving the lines in both escaped string and the output once it is processed through literal_eval.